python/ray/train/v2/_internal/execution/worker_group/worker_group.py

```diff
@@ -668,12 +573,123 @@ def execute_single(
             (T) The output of func.
 
         """
-        self._assert_workers_started()
-
         return ray.get(self.execute_single_async(rank, fn, *fn_args, **fn_kwargs))
 
-    def __len__(self) -> int:
-        return len(self._workers)
+    #####################################################################################
+    # Utility Methods
+    #####################################################################################
+
+    def has_started(self) -> bool:
+        return self._worker_group_state is not None
 
     def get_workers(self) -> List[Worker]:
-        return self._workers
+        # TODO: Access workers through WorkerGroupState instead?
+        if not self._worker_group_state:
+            return []
+        return self._worker_group_state.workers
+
+    def get_worker_group_context(self) -> Optional[WorkerGroupContext]:
+        return self._worker_group_context
+
+    def get_worker_group_state(self) -> Optional[WorkerGroupState]:
+        # TODO: Is this needed?
+        return self._worker_group_state
+
+    def _assert_worker_group_started(self):
+        if not self.has_started():
+            raise ValueError("Worker group not started.")
+
+    def __len__(self) -> int:
+        # TODO: Should this be implemented in WorkerGroupState only?
+        return len(self.get_workers())
+
+    #########################################################################################
+    # Static Utility Methods
+    #########################################################################################
+
+    @staticmethod
+    def _assign_worker_ranks(workers: List[Worker]) -> List[Worker]:
+        """Assign world ranks to workers by increasing node id and GPU id.
+
+        Initializes the `DistributedContext` for each worker.
+
+        Returns:
+            workers: Workers sorted by increasing world rank,
+                with the `DistributedContext` set.
+        """
+        workers = WorkerGroup._sort_workers_by_node_id_and_gpu_id(workers)
+
+        node_ip_to_workers = collections.defaultdict(list)
+        for worker in workers:
+            node_ip_to_workers[worker.metadata.node_ip].append(worker)
+        node_ips = list(node_ip_to_workers.keys())
+
+        for world_rank, worker in enumerate(workers):
+            distributed_context = DistributedContext(
+                local_rank=node_ip_to_workers[worker.metadata.node_ip].index(worker),
+                local_world_size=len(node_ip_to_workers[worker.metadata.node_ip]),
+                world_rank=world_rank,
+                world_size=len(workers),
+                node_rank=node_ips.index(worker.metadata.node_ip),
+            )
+            worker.distributed_context = distributed_context
+
+        return workers
+
+    @staticmethod
+    def _sort_workers_by_node_id_and_gpu_id(
```

Comment:
Made these static because they're just generic utilities that don't need/modify any `WorkerGroup` state